### PR TITLE
Fix storage layout mismatch bugs

### PR DESCRIPTION
**Pull request by SuperN1ck**
Hey @brentyi 
this PR should fix: https://github.com/brentyi/jaxfg/issues/12
Not sure about sorting the full list of `variables` as `sorted` is not in-place afaik. If performance is an issue, we could scan over the full list first and extract all unique `type(variable)`-keys and sort them rather the full list.
Let me know what you think :)

### CODE REVIEW

**Review by brentyi — COMMENTED** [review]
Thanks @SuperN1ck! And @AvanDavad for catching this; this is a major bug. There used to be a sort here, but seems like I dropped it in f2165307c77dda07e88237149e2dc47f43cbea06.

On the suggested fix: sorting by type is likely not enough, because specific variables within each type may be shuffled. Also slightly concerned with sorting by the name of each type, because hypothetically there could be unique types defined in different modules that happen to share the same name.

Two things for moving forward (I can help with this):
- Unit test would be nice :1st_place_medal:
- Rather than sorting all of the variables passed into `StorageMetadata` to make the ordering consistent, perhaps we could actually accept "shuffled" variable assignments as input to methods like `solve()`, which can be unshuffled at the start of the call and then re-shuffled if another assignments object is returned; this was my original intention but I think I just... forgot to actually implement it.

Does this make sense @SuperN1ck?

**Comment by SuperN1ck** [comment]
All valid points @brentyi. I just added a basic test, feel free to modify it to your style.

And yes, the sorting is definitely not optimal. I think the root of the problem lies in the two array accesses:
`assignments.storage[indices]` in `_factor_stack.py` 
https://github.com/brentyi/jaxfg/blob/f5204945bb6afa444810e6163e9a913bdbdd636b/jaxfg/core/_factor_stack.py#L166
https://github.com/brentyi/jaxfg/blob/f5204945bb6afa444810e6163e9a913bdbdd636b/jaxfg/core/_factor_stack.py#L189
If I am not mistaken? Basically, here, the pre-calculated `indices` depend on the `StorageMetadata` layout when the factor stack is created but `assignments.storage` depends on the layout when the `VariableAssignments` is created. The sorting of the variables ensures that the layout is the same.

Related to your suggestion to allow arbitrary sorted value assignments. An immediate, obvious and different fix would be calling `assignments.get_value(variable)` rather than directly accessing `storage`, but this is not possible as the variables in `self.factor.variables` are anonymized. This anonymization would also hinders us from your suggested unshuffling of the assignment object as we do not know how they were initially shuffled when the factor stack was created. But please correct me if I am wrong or overlook something, you have a much better overview of the code!

I think there are two possible ways to go around this:
1. We come up with a better canonical way to sort variables before creating the underlying `storage_metadata`. As currently, this is then used when constructing the `storage_metadata` within the variable assignment and the stacked factor graph respectively and thus, array indices are in sync
2. We modify the VariableAssignment constructors to not take in the variables but rather a `stacked_factor_graph` from which we can retrieve the `storage_metadata` object used for the underlying graph. This would break the current API though. As a nice side-effect we would also ensure that the variable assignments are always consists of all variables present in the factor graph. E.g. currently, when doing something like this
```
variables = [Variable_A(), Variable_A()]
VariableAssignments.make_from_dict({variables[0]: <...>}) # No variables[1]! 
...
```
will silently work fine but shouldn't or at least throw a warning? For context, the assignment created will have `storage.shape = (1,)`. The variable indices created when stacking the factors though would be something like `[[0], [1]]` (depending on graph structure bla). Now the call mentioned above
`assignments.storage[indices]` wouldn't work in normal numpy but does in [JAX](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#out-of-bounds-indexing). Passing the stacked graph to the `VariableAssignments.make_from_...` would prevent missing indices and the indices misalignments.

**Comment by brentyi** [comment]
Appreciate the test + detailed note. :)

`get_value()` is certainly a bit messy to use because it only lets us pull out a single variable value at a time; indexing is much nicer for performance.

Can you take a look at the code I just pushed? It fixes the underlying issue you pointed out and I *think* addresses the other concerns you had. I agree in particular that handling the reordering within the `FactorStack` implementation would be challenging, but the key thing is that we can just slap an assert there and handle reordering at a higher level where it's easier (for example, at the start and end of nonlinear solves). 

In contrast to (1), the current approach keeps the variable ordering more predictable by just matching whatever's passed into the `VariableAssignments` instantiation (important for `get_stacked_value()`), and in contrast to (2) doesn't break any existing APIs (although I see the appeal of something like `graph.make_assignments(...)`).

File: jaxfg/core/_variable_assignments.py [hunk 1]
```diff
@@ -73,13 +74,54 @@ def make_from_partial_dict(
...
+
+        The primary motivation of this method is that the storage layout of an
+        assignments object can sometimes be shuffled with respect to the layout
+        expected by a graph (StackedFactorGraph)."""
+
+        # No-op if storage layouts already match.
+        # if self.storage_layout == storage_layout:
+        #     return self
+
+        assert self.storage_layout.dim == storage_layout.dim
```
Comment (SuperN1ck) [inline jaxfg/core/_variable_assignments.py:100]: If I am not mistaken, these asserts address my concern when passing an incomplete initial assignment, as it should throw a `KeyError`? I am fine with this solution.

File: jaxfg/core/_variable_assignments.py [hunk 2]
```diff
@@ -73,13 +74,54 @@ def make_from_partial_dict(
...
             ],
             axis=0,
         )
-        assert storage.shape == (storage_metadata.dim,)
+        assert storage.shape == (storage_layout.dim,)
 
-        return VariableAssignments(storage=storage, storage_metadata=storage_metadata)
+        return VariableAssignments(storage=storage, storage_layout=storage_layout)
+
+    @functools.partial(jax.jit, static_argnums=1)
```
Comment (SuperN1ck) [inline jaxfg/core/_variable_assignments.py:85]: I am unsure about the jitting here and when this gets recompiled. 

What I am thinking about: The argument `initial_assignments` for `solve` is not static (which makes sense). In `solve` itself we then call this function as
```
assignments = initial_assignments.update_storage_layout(graph.storage_layout)
```
In this call the first argument (`self`) is not static but the second is static, so this is fine to compile.
For the second call in `solve`
```
state.assignments.update_storage_layout(initial_assignments.storage_layout)
```
`initial_assignments.storage_layout` is a non-static variable (as passed to `solve`), but now gets passed as static to `update_storage_layout` so this should trigger a re-jitting if we pass a different `initial_assignments`-object to `solve`? 

If I am not mistaken in `update_storage_layout` the second argument `storage_layout` is static to allow iteration over `variables` with the `for`-loop. To allow both arguments to be non-static, could we explicitly pass a third argument `variables` and just make this static?

**Review by SuperN1ck — COMMENTED** [review]
I think this implementation is great without breaking anything and keeps everything under the hood!
What's your opinion on `jit`-ing?

File: jaxfg/core/_variable_assignments.py (hunk 1 above)
Comment (brentyi) [reply jaxfg/core/_variable_assignments.py:100]: That's right!

File: jaxfg/core/_variable_assignments.py (hunk 2 above)
Comment (brentyi) [reply jaxfg/core/_variable_assignments.py:85]: This is a good question, and I agree it's a bit confusing.

Note that the `StorageLayout` class is currently a regular `dataclasses.dataclass` and not a `jdc.pytree_dataclass`, and should always be marked as static under JIT. `initial_assignments` isn't static, but if you check the implementation of `VariableAssignments`, `VariableAssignments.storage_layout` should be marked as a static field (via `jdc.static_field()`). So in all of the cases, changing the storage layout or variables will re-trigger JIT compilation.

It's possible in the future that we can make `storage_layout` not static, but there will be a runtime hit from making the storage indexing dynamic and the utility is not entirely clear to me.

**Review by brentyi — APPROVED** [review]
